gateway.go

```diff
@@ -112,7 +115,19 @@ func (api *bifrostGateway) GetBlock(ctx context.Context, c cid.Cid) (blocks.Bloc
 }
 
 func (api *bifrostGateway) GetIPNSRecord(ctx context.Context, c cid.Cid) ([]byte, error) {
-	return api.routing.GetValue(ctx, "/ipns/"+c.String())
+	// Fails fast if the CID is not an encoded Libp2p Key, avoids wasteful
+	// round trips to the remote routing provider.
+	if mc.Code(c.Type()) != mc.Libp2pKey {
+		return nil, errors.New("provided cid is not an encoded libp2p key")
+	}
+
+	// The value store expects the key itself to be encoded as a multihash.
+	id, err := peer.FromCid(c)
+	if err != nil {
+		return nil, err
+	}
+
+	return api.routing.GetValue(ctx, "/ipns/"+string(id))
```

Comment:
@lidel I think the problem here is that I used 0.19 to test and the wire format changed between 0.18 and 0.19 - I will fix this as soon as I can. The problem likely lies in `routing.go`.

Reply:
@lidel after testing this locally, I think the problem is that you're passing the gateway to `--kubo-rpc`. Kubo exposes a basic version of the API on the gateway port via GET requests. We need those to redirect (#13).

However, `/api/v0/routing/get` is only exposed on the API port via POST. So then you'd need to give `--kubo-rpc` your API URL. But then the redirects won't work anymore.

We can:

1. Have two flags: one for the redirect RPC (that uses GET, aka, the gateway) and one for the RPC that provides `/routing/get` via POST.
2. Go to Kubo and make `/routing/get` available via GET on the gateway API. But then the gateways would need to upgrade.

I don't think there's a good way of testing this to be honest.

Reply:
TODO: support both 0.18 and 0.19 .

Reply:
Filled https://github.com/ipfs/kubo/issues/9638  to fixup wire format for `/api/v0/routing/get`

Reply:
@lidel  routing/get wire format fixed in https://github.com/ipfs/kubo/pull/9639. Pushed a change here with the fix. It will support both 0.18 and 0.19 with the fix linked.

Reply:
> Our reliance on Kubo RPC for finding IPNS record is (hopefully) temporary, we should keep code related to it minimal.

Imagine using `?format=ipns-record` 😛